source/agora/consensus/EnrollmentManager.d

```diff
@@ -773,6 +773,25 @@ public class EnrollmentManager
         return this.validator_set.addPreimage(preimage);
     }
 
+    /***************************************************************************
+
+        Add pre-images information for enrolled validators
+
+        Params:
+            preimage = the pre-images information to add
+
+    ***************************************************************************/
+
+    public void addPreimages (const PreImageInfo[] preimages) @safe nothrow
+    {
+        foreach (image; preimages)
+        {
+            this.validator_set.addPreimage(image);
```

Comment:
You're ignoring the return value of `addPreimage`. What if a node sends you the wrong preimages that don't validate? Then you should exit early and try getting them from another node.

Specifically another node could send you a giant array of `PreImageInfo`. It would be good to return early if any of the preimages are invalid.

Reply:
Right, it would be better.
Will fix it.